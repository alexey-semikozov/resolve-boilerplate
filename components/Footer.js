import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Link from 'next/link';

function renderTodoCount(activeCount) {
  const itemWord = activeCount === 1 ? 'item' : 'items';

  return (
    <span className="todo-count">
      <strong>{activeCount || 'No'}</strong> {itemWord} left
    </span>
  );
}

function renderFilterLink({ filter, titles, selectedFilter }) {
  return (
    <Link href={`/?filter=${filter}`} as={`/${filter}`}>
      <a
        className={classnames({ selected: filter === selectedFilter })}
        style={{ cursor: 'pointer' }}
      >
        {titles[filter]}
      </a>
    </Link>
  );
}

function Footer(props) {
  return (
    <footer className="footer">
      {renderTodoCount(props.activeCount)}
      <ul className="filters">
        {Object.keys(props.titles).map(filter => (
          <li key={filter}>
            {renderFilterLink({
              filter,
              titles: props.titles,
              selectedFilter: props.filter
            })}
          </li>
        ))}
      </ul>
    </footer>
  );
}

Footer.propTypes = {
  completedCount: PropTypes.number.isRequired,
  activeCount: PropTypes.number.isRequired,
  filter: PropTypes.string.isRequired,
  titles: PropTypes.object.isRequired
};

export default Footer;
