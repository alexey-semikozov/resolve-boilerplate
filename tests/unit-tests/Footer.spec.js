import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';
import Footer from '../../components/Footer';
import {
  SHOW_ALL,
  SHOW_ACTIVE,
  SHOW_COMPLETED
} from '../../components/MainSection';

const setup = propOverrides => {
  const props = Object.assign(
    {
      completedCount: 0,
      activeCount: 0,
      filter: SHOW_ALL,
      onClearCompleted: jest.fn(),
      onShow: jest.fn(),
      titles: {
        [SHOW_ALL]: 'All',
        [SHOW_ACTIVE]: 'Active',
        [SHOW_COMPLETED]: 'Completed'
      }
    },
    propOverrides
  );

  const shallowrenderer = new ReactShallowRenderer();
  shallowrenderer.render(<Footer {...props} />);
  const output = shallowrenderer.getRenderOutput();

  const component = renderer.create(<Footer {...props} />);
  const tree = component.toJSON();

  return {
    props,
    output,
    tree
  };
};

const getTextContent = elem => {
  const children = Array.isArray(elem.props.children)
    ? elem.props.children
    : [elem.props.children];

  return children.reduce(
    (out, child) =>
      // Concatenate the text
      // Children are either elements or text strings
      out + (child.props ? getTextContent(child) : child),
    ''
  );
};

describe('components', () => {
  describe('Footer', () => {
    it('should render container', () => {
      const { tree } = setup();
      expect(tree).toMatchSnapshot();
    });

    it('should display active count when 0', () => {
      const { output } = setup({ activeCount: 0 });
      const [count] = output.props.children;
      expect(getTextContent(count)).toBe('No items left');
    });

    it('should display active count when above 0', () => {
      const { output } = setup({ activeCount: 1 });
      const [count] = output.props.children;
      expect(getTextContent(count)).toBe('1 item left');
    });
  });
});
