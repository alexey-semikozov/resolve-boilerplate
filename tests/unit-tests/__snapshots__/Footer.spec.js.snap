// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Footer should render container 1`] = `
<footer
  className="footer"
>
  <span
    className="todo-count"
  >
    <strong>
      No
    </strong>
     
    items
     left
  </span>
  <ul
    className="filters"
  >
    <li>
      <a
        className="selected"
        href="/all"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        All
      </a>
    </li>
    <li>
      <a
        className=""
        href="/active"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        Active
      </a>
    </li>
    <li>
      <a
        className=""
        href="/completed"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        Completed
      </a>
    </li>
  </ul>
</footer>
`;
